### PR TITLE
fix: Fix assert when calling a setter property

### PR DESCRIPTION
**Pull request by saulecabrera**
Emits an error diagnostic when calling a setter property. This case should be treated in the same way getters are treated. The issue is reproducible with the following code:

```ts
class B {
  set prop(a: i32) {}
}

export function add(): void {
  // Actual: assert
  // Expected: error diagnostic specifying that this expression is not callable. 
  (new B()).prop(1);
}
```

I emitted the diagnostic early, I'm assuming that there's no point in compiling the expression when this case is met as the setter expression is compiled when an assignment expression is found (might be wrong about this).

- [x] I've read the contributing guidelines

### CODE REVIEW

**Comment by MaxGraey** [comment]
I guess need the same for getter as well?

**Comment by dcodeIO** [comment]
Iirc it is allowed to *call* a getter as a function, and the error seen here is rather because of the property not having a getter, not because it has a setter.

**Comment by MaxGraey** [comment]
I mean this
```ts
class B {
  set prop(a: i32) {}
  get prop(): i32 { return 1 }
}

export function add(): void {
  (new B()).prop(1); // should be error
  (new B()).prop();  // should be error as well
}
```

**Comment by saulecabrera** [comment]
> I mean this
> 
> ```ts
> class B {
>   set prop(a: i32) {}
>   get prop(): i32 { return 1 }
> }
> 
> export function add(): void {
>   (new B()).prop(1); // should be error
>   (new B()).prop();  // should be error as well
> }
> ```

It already does

**Comment by saulecabrera** [comment]
@dcodeIO 

> because of the property not having a getter (which is unexpected), not because it has a setter.

Having only either or is valid AFAIK. But I see your point, this is that part that I was a bit unsure about, this should error only if there's a setter and no getter.

**Comment by saulecabrera** [comment]
Just verified, and you were right:

- When both set and get are present, the diagnostics are emitted correctly
- When only the getter is present, the diagnostic is emitted correctly
- When only the setter is present, the crash occurs.

I've amended the change to fix only the last condition.

File: src/compiler.ts [hunk 1]
```diff
@@ -6335,6 +6335,15 @@ export class Compiler extends DiagnosticEmitter {
...
         let propertyInstance = <Property>target;
+
+        if (propertyInstance.setterInstance && !propertyInstance.getterInstance) {
+          this.error(
+            DiagnosticCode.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature_Type_0_has_no_compatible_call_signatures,
+            expression.range, this.currentType.toString()
+          );
+          return module.unreachable();
+        }
+
```
Comment (dcodeIO) [inline src/compiler.ts:6346]: Seems like only the presence of a getter is relevant here (would not call a setter anyway), so this can perhaps be simplified to
```ts
let getterInstance = propertyInstance.getterInstance;
if (!getterInstance) {
  this.error(
    DiagnosticCode.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature_Type_0_has_no_compatible_call_signatures,
    expression.range, this.currentType.toString()
  );
  return module.unreachable();
}
```
i.e. replacing the `assert` with an `if`.

Comment (dcodeIO) [reply src/compiler.ts:6346]: Also not sure about `this.currentType` here. Seems like a case for `getTypeOfElement` on `target` rather, which may return `null`, so if it does better printing a "is not represented by a type" kind of error.

Comment (saulecabrera) [reply src/compiler.ts:6346]: Yeah, this works too.

Comment (dcodeIO) [reply src/compiler.ts]: I think my comment on `this.currentType` may have got lost somehow (no problem), or is this not applicable?

Comment (saulecabrera) [reply src/compiler.ts]: > I think my comment on this.currentType may have got lost somehow (no problem), or is this not applicable?

After looking at `compileExpression` indeed, it sets `this.currentType` to the contextual type passed in, which ensures that the diagnostic reported after (around line 6359) has the accurate type. 

So yeah I think that calling `resolver.getTypeOfElement(target)` is the safest route. 

>  if it does better printing a "is not represented by a type" kind of error.

Just to make sure that we are on the same page, with this, do you mean using `DiagnosticCode.Expression_cannot_be_represented_by_a_type`

Comment (saulecabrera) [reply src/compiler.ts]: Before pushing the change, there's something that remains unclear to me, since properties can only belong to a class and properties are (always?) typed elements (according to the TypedElement class), in which case is the latter diagnostic code going to be useful?

Comment (dcodeIO) [reply src/compiler.ts]: Hmm, if that's the case, an `assert(getTypeOfElement(...))` may be sufficient :)

Comment (saulecabrera) [reply src/compiler.ts]: I'll go with your latest suggestion, if that `assert` is somehow surfaced I'll be glad to fix it (and also curious to know why it happened).

Comment (saulecabrera) [reply src/compiler.ts]: I found a counterexample that hints that the usage of `getTypeOfElement` is not the way to go and instead we should keep `this.currentType`. 

`getTypeOfElement` will work as long as the type of the element is not void https://github.com/AssemblyScript/assemblyscript/blob/master/src/resolver.ts#L880, but one could do 


```ts
class B {
  set prop(a: void) {}
}
```

which will result in an assertion coming from the line linked above. 

Using `this.currentType`  doesn't hit the assert and instead, in this case, the two expected diagnostics are returned:

```
ERROR TS1110: Type expected.

    set prop(val: void) {}
                  ~~~~
 in assembly/index.ts(8,18)

ERROR TS2349: Cannot invoke an expression whose type lacks a call signature. Type 'void' has no compatible call signatures.

    (new B()).prop(1);
    ~~~~~~~~~~~~~~~~~
 in assembly/index.ts(12,4)
```
I _think_ that this is because the current type is correctly resolved here https://github.com/AssemblyScript/assemblyscript/blob/master/src/compiler.ts#L6330-L6335

TLDR: according to this example, `getTypeOfElement` is not applicable.

Comment (saulecabrera) [reply src/compiler.ts]: This behaviour is also the same if you have `void` as the type of function parameter, which makes sense since internally setters and getters are resolved as functions

Comment (saulecabrera) [reply src/compiler.ts]: @dcodeIO taking into account my latest comment, is there anything else that you think is worth exploring here? If not I believe this is ready to go.

Comment (dcodeIO) [reply src/compiler.ts]: I have been thinking about the implications of removing the `assert` in `getTypeOfElement`, but haven't yet thought it through. Like, an error of "void is not callable" looks fine, if that's indeed the type, but it may affect other code paths in some way.

Comment (saulecabrera) [reply src/compiler.ts]: Yeah, good point, FWIW it was introduced here https://github.com/AssemblyScript/assemblyscript/pull/864/files, and removing the assert and re-running the tests locally doesn't break anything.

Comment (saulecabrera) [reply src/compiler.ts]: After digging deeper into this, I believe I finally found the root of the `void` assertion. 

First of all, I couldn't track down a case in which the type here would be void, effectively all `typedElements` which can be any of: `EnumValue`, `Property`, `Global`, `Field`, `Local`, `Index Signature`, `Type definition`, `Function`, `Class` and `Enum` are properly resolved when they this method is called; with the linked PR above, there was an exception with lazy globals, in which if the global wasn't typed (for whatever reason) the expression wouldn't be resolved, potentially leaving the `type` property of the `TypedElement` class as void, but that case was correctly fixed by the PR above. I'm my latest change I've removed the `void` assertion and returned to use `resolver#getTypeOfElement`. 

---

@MaxGraey do you have any kind of context in this part of the compiler so that you can review?

Comment (MaxGraey) [reply src/compiler.ts]: > @MaxGraey do you have any kind of context in this part of the compiler so that you can review?

No, so I'll need time to investigate to related parts

Comment (saulecabrera) [reply src/compiler.ts]: Makes sense, thanks!

Comment (MaxGraey) [reply src/compiler.ts]: @saulecabrera So it fix issue (assert) occurring only when the setter is present and add more proper diagnostic error. It looks legit for me.
